variable.c

```diff
@@ -1440,7 +1441,12 @@ rb_obj_ivar_set(VALUE obj, ID id, VALUE val)
 
         RUBY_ASSERT(!rb_shape_obj_too_complex(obj));
 
-        if (UNLIKELY(shape->next_iv_index >= num_iv)) {
+        VALUE klass = rb_obj_class(obj);
+        bool too_complex = RCLASS_EXT(klass)->variation_count >= SHAPE_MAX_VARIATIONS;
```

Comment:
Can we move the check `RCLASS_EXT(rb_obj_class(obj))->variation_count >= SHAPE_MAX_VARIATIONS` to a separate function so that we don't have duplicate logic here and `rb_shape_get_next`?